apiserver/facades/agent/uniter/networkinfocaas.go

```diff
@@ -0,0 +1,208 @@
+	}
+	corenetwork.SortAddresses(addrs)
+
+	// We record the interface addresses as the machine local ones - these
+	// are used later as the binding addresses.
+	// For CAAS models, we need to default ingress addresses to all available
+	// addresses so record those in the default ingress address slice.
+	var interfaceAddr []network.InterfaceAddress
+	var defaultIngressAddresses []string
+	for _, a := range addrs {
+		if a.Scope == corenetwork.ScopeMachineLocal {
+			interfaceAddr = append(interfaceAddr, network.InterfaceAddress{Address: a.Value})
+		} else {
+			defaultIngressAddresses = append(defaultIngressAddresses, a.Value)
+		}
+	}
+
+	networkInfos := make(map[string]machineNetworkInfoResult)
+	networkInfos[corenetwork.AlphaSpaceId] = machineNetworkInfoResult{
+		NetworkInfos: []network.NetworkInfo{{Addresses: interfaceAddr}},
+	}
+
+	for endpoint, space := range bindings {
+		// The binding address information based on link layer devices.
+		info := machineNetworkInfoResultToNetworkInfoResult(networkInfos[space])
+
+		// Set egress and ingress address information.
+		info.EgressSubnets = endpointEgressSubnets[endpoint]
+
+		ingressAddrs := make([]string, len(endpointIngressAddresses[endpoint]))
+		for i, addr := range endpointIngressAddresses[endpoint] {
+			ingressAddrs[i] = addr.Value
+		}
+		info.IngressAddresses = ingressAddrs
+
+		// If there is no ingress address explicitly defined for a given
+		// binding, set the ingress addresses to either any defaults set above,
+		// or the binding addresses.
+		if len(info.IngressAddresses) == 0 {
+			info.IngressAddresses = defaultIngressAddresses
+		}
+
+		if len(info.IngressAddresses) == 0 {
+			ingress := spaceAddressesFromNetworkInfo(networkInfos[space].NetworkInfos)
+			corenetwork.SortAddresses(ingress)
+			info.IngressAddresses = make([]string, len(ingress))
+			for i, addr := range ingress {
+				info.IngressAddresses[i] = addr.Value
+			}
+		}
+
+		// If there is no egress subnet explicitly defined for a given binding,
+		// default to the first ingress address. This matches the behaviour when
+		// there's a relation in place.
+		if len(info.EgressSubnets) == 0 && len(info.IngressAddresses) > 0 {
+			var err error
+			info.EgressSubnets, err = network.FormatAsCIDR([]string{info.IngressAddresses[0]})
+			if err != nil {
+				return result, errors.Trace(err)
+			}
+		}
+
+		result.Results[endpoint] = info
+	}
+
+	return dedupNetworkInfoResults(result), nil
+}
+
+// getRelationNetworkInfo returns the endpoint name, network space
+// and ingress/egress addresses for the input relation ID.
+func (n *NetworkInfoCAAS) getRelationNetworkInfo(
+	relationId int,
+) (string, string, corenetwork.SpaceAddresses, []string, error) {
+	rel, endpoint, err := n.getRelationAndEndpointName(relationId)
+	if err != nil {
+		return "", "", nil, nil, errors.Trace(err)
+	}
+
+	cfg, err := n.app.ApplicationConfig()
+	if err != nil {
+		return "", "", nil, nil, errors.Trace(err)
+	}
+
+	var pollAddr bool
```

Comment:
@ycliuhw @wallyworld I want to push this into the local `NetworksForRelation` method, because that would remove the boolean argument (it's already moot for IAAS).

What concerns me is that in `EnterScope`, `NetworksForRelation` is called directly and always passed `false`. Is that actually correct? I would have thought we want to poll for an address based on the `ServiceType`.

Reply:
@manadart the address info is expected to come from a network-get call. Juju used to not have that and used to seed the relation data directly from EnterScope (back when things were a lot simpler and IAAS units always had addresses available). For backwards compatibility, we continue to add whatever address info we know at the time of EnterScope(). Specifically for k8s, the address info may not be known at that time. But the expectation for "new" charms, charms written since network-get became a thing, is to always use that tool to get what they need and explicitly update relation data. In these cases, pollAddr is not hard coded to false.